[2단계 - 자동차 경주 리팩토링] 토리(천은정) 미션 제출합니다.

ezzanzzan opened the pull request:
안녕하세요 찰리 ! 🙇‍♀️

1단계 미션 merge 후 남겨주신 피드백은 다 확인하였는데,
아직 모르는 부분이 많아 피드백 주신 부분에 대해 정확한 답을 내리지 못했습니다 😢 
고민하는 과정에서 시간 분배에 실패해서 수정하지 못한 부분이 많습니다 .. 

피드백 주신 내용에 대해서는 더 고민해본 뒤 이해하고 마저 리팩토링하며 코멘트에 남기겠습니다 💪

오늘도 좋은 하루 되세요 🫶

ezzanzzan committed:
test: 전진 테스트 ParameterizedTest name 수정

ezzanzzan committed:
test: 메서드 이름 네이밍 컨벤션 적용

ezzanzzan committed:
test: 전진 테스트 메서드 인자 이름 변경

ezzanzzan committed:
refactor: run 메소드 역할 변경

ezzanzzan committed:
refactor: random 객체 생성 위치 변경

ezzanzzan committed:
refactor: validateNameInput 메서드 이름 변경

ezzanzzan committed:
refactor: printCarResult 메소드 내의 for문 repeat()로 변경

ezzanzzan committed:
refactor: LinkedHashMap interface인 Map을 반환하도록 변경

ezzanzzan committed:
refactor: constant 패키지와 접근 제어자 변경

ezzanzzan committed:
refactor: entrySet을 사용해서 map의 내용물을 꺼내도록 수정

Gomding commented on src/main/java/racingcar/controller/RacingcarController.java:
```diff
@@ -39,16 +45,16 @@ public void initializeTryCount() {
         OutputView.printResultMessage();
         for (int i = 0; i < tryCount; i++) {
-            LinkedHashMap<String, Integer> roundResult = carService.runRound();
+            Map<String, Integer> roundResult = carService.runRound();
```
1단계 리뷰때 물어보지못했는데, Map을 사용해서 OutputView 에 넘겨주기로 결정한 이유가 궁금해요!

ezzanzzan replied:
mvc 패턴에서 view와 model(domain) 사이에 의존이 있어서는 안된다는 생각에 Map을 사용해서 OutputView에 넘겨주게 된 것 같습니다.. 
```
private void playRound(Cars cars, TryCount tryCount) {
        for (int i = 0; i < tryCount.getTryCount(); i++) {
            List<Car> roundResult = cars.runRound();
            outputView.printRoundResult(roundResult);
        }
        outputView.printRoundResult(cars.getCars());
    }
```
리팩토링 과정에서 Map을 사용해서 넘겨주게 될 때 변환하는 코드를 쓰는 것이 더 복잡해지는 것 같아
List<Car> 형태로 OutputView에 넘겨주는 방식으로 변경하였습니다.

Gomding commented on src/main/java/racingcar/domain/Car.java:
```diff
@@ -15,8 +14,8 @@ public Car(String name, int distance) {
-        if (name.length() > CarConstant.NAME_MAX_LENGTH.getNumber()
-                || name.length() <= CarConstant.NAME_MIN_LENGTH.getNumber()) {
+        if (name.length() > NAME_MAX_LENGTH
+                || name.length() <= NAME_MIN_LENGTH) {
```
토리의 개행 기준이 궁금해요!

이미 알고계실수도 있지만 IntelliJ의 코드를 작성하고 나서 Reformat Code 기능을 항상 실행해보는걸 추천드립니다!
단축키는 맥북기준 command + option + L / 윈도우기운 Ctrl + Alt + L 입니다 😄 
패키지 단위, 파일 단위로도 실행이 가능합니다 :)

코드의 Format을 계속 통일시켜주면 불필요한 코드가 제거되고, 코드가 익숙한 형태로 지속가능한데요. 이를 통해서 코드 읽는 사람이 예측가능한 형태를 만들 수 있다고 생각해요 :)

ezzanzzan replied:
기존에 코드를 구현할 때는 개행 기준에 대해 생각해보지 않고, 자동 정렬을 통해서 정렬을 해왔습니다. 다만 이번 리팩토링 과정에서 stream을 사용하게 되면서 어떤 기준을 가지고 개행을 해야할까 생각해보게 되었습니다 !
stream을 사용할 때는 .을 기준으로 한 줄에 하나의 기능을 가져 가독성을 높이는 것을 목표로 개행 기준을 정했습니다 ˙ᵕ˙ 
다른 경우에서는 글자수를 정해두고 한 줄에 해당 글자 수가 넘어간다면 **한 줄에 하나의 기준**이 되도록 개행을 할 수 있도록 글자수의 기준에 대해 코드를 구현하며 계속 고민해봐야 할 것 같습니다  !!

Gomding commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -37,7 +37,7 @@ class TestNameFormat {
         @ParameterizedTest(name = "{displayName} {index} ==> name : ''{0}''")
         @ValueSource(strings = {" ", "abc!", "123* "})
         @DisplayName("자동차 이름이 영숫자가 아닌 경우 예외 발생")
```
이부분은 이전단계에서 피드백을 드렸어야했는데, 현재 PR에는 코멘트는 남길 수 없어 우선 이곳에 남겨봅니다 ㅠㅠ

```java
    @Nested
    @DisplayName("자동차 이름 형식 테스트")
    class TestNameFormat {
```
이건 피드백이 아닌 단순 의견이 궁금해서 물어보는 것입니다!
`@Nested` 애너테이션을 통해 테스트 컨텍스트를 나눠주셨는데, 컨텍스트를 `자동차 객체 생성 테스트` 라는 단위로 묶은것이 아닌 Validation 단위로 묶은 배경이 궁금해요! 😄

ezzanzzan replied:
Nested를 사용하며 각 테스트마다 성공,실패의 테스트를 진행하고 이것을 하나의 테스트로 묶고자 해서 컨텍스트를 validation 단위로 묶게 되었습니다 !

Gomding commented on src/main/java/racingcar/service/CarService.java:
```diff
@@ -3,31 +3,34 @@
     private void makeCar(List<String> carNames) {
         for (String carName : carNames) {
-            Car car = new Car(carName, CarConstant.DISTANCE_DEFAULT.getNumber());
+            Car car = new Car(carName, DISTANCE_DEFAULT);
```
Car 객체를 생성할 때, Car 생성자 내부에서 DISTANCE_DEFAULT 로 초기화를 시켜보는건 어떨까요?

Car 클래스에 주 생성자와 부 생성자를 활용해볼 수 있겠어요!
https://programmer-ririhan.tistory.com/376

ezzanzzan replied:
조언해주신 방법인 생성자 체이닝을 통해
```
    public Car(String name) {
        this(name, DISTANCE_DEFAULT);
    }

    public Car(String name, int distance) {
        this.name = new Name(name);
        this.distance = new Distance(distance);
    }
```
위처럼 리팩토링 해보았습니다 ˙ᵕ˙ !
덕분에 코드가 좀 더 가독성 있고 깔끔해질 수 있었습니다 🥺

Gomding commented on src/main/java/racingcar/service/CarService.java:
```diff
@@ -36,10 +39,10 @@ public List<String> splitCarNames(String carNames) {
         for (Car car : cars.getCarInformation()) {
-            Random random = new Random();
             int randomNumber = random.nextInt(RANDOM_NUM_MAX_VALUE);
             runForward(car, randomNumber);
```
Car 객체에게 메시지를 던져서, Car 자신이 전진을 위한 값을 판단하고, 전진할 수 있으면 전진시키도록 변경해보는건 어떤가요? 😄 

```suggestion
            car.runForward(randomNumber)
```

ezzanzzan replied:
리팩토링 과정에서 CarService.java가 Car.java가 가져야하는 책임을 다소 가지고 있었다는 것을 알게되었습니다. 
**RacingcarController**에서 **Cars**에게 메시지를 던지고, **Cars**에서 **Car**로 **랜덤값과 함께 메세지를 던져** Car 에서 스스로 값을 판단하고 전진 할 수 있도록 리팩토링해보았습니다 🥺

Gomding commented on src/main/java/racingcar/service/CarService.java:
```diff
@@ -3,31 +3,34 @@
+    private static final int TRY_COUNT_MIN = 0;
 
-    public void validateNameInput(String carNameInput) {
+    public void makeCars(String carNameInput) {
```
메서드 이름을 실행하는 책임에 맞게 잘 변경해주셨네요 💯

Gomding commented on src/main/java/racingcar/service/CarService.java:
```diff
@@ -3,31 +3,34 @@
 
-    public void validateNameInput(String carNameInput) {
+    public void makeCars(String carNameInput) {
         cars = new Cars(new ArrayList<Car>());
```
Cars 라는 일급컬렉션에서 가지는 값은, 생성자에서 바로 초기화를 시켜주면 어떨까요?

```java
Cars(List<Car> carInformation) {
    this.carInformation = carInformation
}
```

아래는 일급 컬렉션에 대한 좋은글이 있어 공유드립니다!
https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/

ezzanzzan replied:
makeCars 메소드 뿐만 아니라 makeCar 메소드도 생성자에서 값을 바로 초기화할 수 있도록 리팩토링 해보았습니다 !
```
    private Cars makeCars() {
        try {
            outputView.printNameInput();
            return new Cars(inputView.readCarNames());
        } catch (IllegalArgumentException e) {
            System.out.println(e.getMessage());
            return makeCars();
        }
    }
```
```
public Cars(List<String> carNames) {
        this.cars = convertToCar(carNames);
    }

    private List<Car> convertToCar(List<String> carNames) {
        return carNames.stream().map(Car::new).collect(Collectors.toList());
    }

```

Gomding reviewed (changes requested):
안녕하세요 토리!
연극 일정에 바쁘셨을텐데 2단계도 진행하느라 정말 고생하셨습니다!
1단계에서 반영이 안된 피드백이 있어서 확인 부탁드려요 :)
추가로 몇 가지 코멘트 드렸습니다. 확인하고 의견 부탁드립니다~

리뷰에 대한 이해가 힘들때는 DM으로 직접 질문주세요~!! 토리가 이해못한 것이 아니라 리뷰를 모호하게 드렸을수도 있으니까요 👀

ezzanzzan committed:
docs: 리팩토링 목록 정의

ezzanzzan committed:
refactor: 자동차 이름을 입력받은 후 split하는 메소드 InputView.java로 위치 변경

ezzanzzan committed:
refactor: 접근 제어자 수정

ezzanzzan committed:
refactor: 불필요한 코드 리팩터링

ezzanzzan committed:
refactor: 접근 제어자 수정

ezzanzzan committed:
refactor: Constants 패키지 변경을 통한 접근 최소화

ezzanzzan committed:
refactor: 불필요한 코드 제거

ezzanzzan committed:
feat: 생성자 체이닝을 통해 객체 생성시 거리 초기화

ezzanzzan committed:
feat: Car의 name, distance VO로 변경

ezzanzzan committed:
feat: Cars 생성자 안에서 List<Car> 생성할 수 있는 기능 구현

- List<String> 형태의 carNames를 받아, convertToCar를 통해 List<Car> 형태로 생성할 수 있는 기능 구현

ezzanzzan committed:
refactor: 메소드 리팩토링

- RacingcarController.java의 run() 메소드 리팩토링
- OutputView.java의 printCarResult(), printWinners() 메소드 for문 stream으로 리팩토링

ezzanzzan committed:
test: 단위 테스트 수정

ezzanzzan committed:
test: tryCount 단위 테스트 기능 구현

ezzanzzan committed:
refactor: 우승자를 찾을 때 winner List 생성 후 반환하도록 리팩토링

Gomding commented on src/main/java/racingcar/domain/TryCount.java:
```diff
@@ -0,0 +1,21 @@
+package racingcar.domain;
+
+public class TryCount {
```
원시값 포장 💯 💯 💯

Gomding commented on src/main/java/racingcar/domain/TryCount.java:
```diff
@@ -0,0 +1,21 @@
+
+public class TryCount {
+    private static final int TRY_COUNT_MIN = 0;
+    private final int tryCount;
```
보통 상수와 객체의 필드는 개행으로 분리시켜 모아두는것이 가독성 측면에서 좋다고 생각해요!

```suggestion
    private static final int A = 0;
    private static final int B = 0;
    private static final int C = 0;
    
    private final int value1;
    private final int value2;
    private final int value3;
```

ezzanzzan replied:
전체적인 클래스를 상수와 객체의 필드를 개행으로 분리하는 방향으로 리팩토링해보았습니다 !

Gomding commented on src/main/java/racingcar/domain/TryCount.java:
```diff
@@ -0,0 +1,21 @@
+
+public class TryCount {
+    private static final int TRY_COUNT_MIN = 0;
+    private final int tryCount;
```
원시값 포장 시 `값의 의미`는 클래스 이름에 표현됐으니 필드는 value 로 표현하는것에 대해서 어떻게 생각하시나요?
토리의 의견이 궁금합니다!

```suggestion
class TryCount {
    ...
    private final int value;

    public int value() {
        return this.value;
    }
}
```

ezzanzzan replied:
처음에는 value보다 값의 의미 자체로 표현하는 게 가독성이 좋다고 생각해 이렇게 사용하게 되었는데, 
밑에 코멘트에도 남겨주신 부분인 getName().getName(), getDistance().getDistance()와 같이 같은 이름의 메서드가 두 번 연속 호출이 되어 저도 어색함을 느끼는 중이었습니다 ㅠㅠ 🥲
메서드를 호출할 때를 생각해보면 찰리께서 추천해주신 방식으로 리팩토링을 하는 것이 더 가독성이 좋을 것 같다고 생각해
TryCount, Name, Distance 클래스의 필드를 모두 value 로 리팩토링하였습니다 ! 😊

Gomding commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -14,8 +14,7 @@ public List<String> readCarNames() {
         try {
-            int tryCount = Integer.parseInt(scanner.nextLine());
-            return tryCount;
+            return Integer.parseInt(scanner.nextLine());
```
불필요한 변수할당을 제거해주셨군요 👍

Gomding commented on src/main/java/racingcar/domain/Distance.java:
```diff
@@ -0,0 +1,17 @@
+
+public class Distance {
+    private int distance;
+
```
Car 에서 distance 를 0으로 자주 사용하게 된다면 미리 생성해두는것도 좋은 방법이 될 수 있습니다 😄 
(현재는 반영하지 않으셔도 괜찮습니다!!)

```suggestion
public class Distance {
    public static final Distance ZERO = new Distance(0);

    private int distance;

}
```

Gomding commented on src/main/java/racingcar/domain/Name.java:
```diff
@@ -0,0 +1,29 @@
+public class Name {
+    private static final int NAME_MAX_LENGTH = 5;
+    private static final int NAME_MIN_LENGTH = 0;
+    private final String name;
```
원시값 포장 👍

Gomding replied:
여기도 개행을 추가해봐도 되겠어요 😄

Gomding commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -1,47 +1,42 @@
+        roundResult
+                .forEach(car -> System.out.println(car.getName().getName()
+                        + CAR_INFORMATION_DELIMITER
+                        + DISTANCE_COMMAND.repeat(car.getDistance().getDistance())));
```
car.getName().getName()
car.getDistance().getDistance()
를 보면 같은 이름의 메서드가 두번 연속 호출되니 어색함이 느껴지네요 🤔 

car.getName().getValue()
car.getDistance().getValue()
같이 변경해보면 어떨까요?

Gomding commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -1,47 +1,42 @@
-        }
+    public void printRoundResult(List<Car> roundResult) {
+        roundResult
+                .forEach(car -> System.out.println(car.getName().getName()
```
.forEach 나 .Stream 과 같이 시작되는 부분은 개행이 없어도 된다고 생각해요!

Gomding commented on README.md:
```diff
@@ -5,3 +5,55 @@
+    - [x] 우승자를 출력하는 기능
+
+---
+### 리팩토링 목록
```
리팩토링 목록 👍

Gomding commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -2,41 +2,54 @@
-    public void addCarInformation(Car car) {
-        this.carInformation.add(car);
+    private List<Car> convertToCar(List<String> carNames) {
+        return carNames.stream().map(Car::new).collect(Collectors.toList());
```
여기서 stream 의 메서드 호출마다 개행 해주셔도 되겠어요 😃

ezzanzzan replied:
stream을 사용할 때 메서드 호출마다 개행을 하는 것으로 기준을 세웠는데, 이 부분을 놓친것같습니다 🥺
```
    public Cars(String carNames) {
        this(Stream.of(carNames.split(DELIMITER))
                .map(Car::new)
                .collect(Collectors.toList()));
    }
```
와 같이 리팩토링 해보았습니다 !! 💪

Gomding commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -2,41 +2,54 @@
-        for (Car car : this.carInformation) {
-            maxDistance = Math.max(car.getDistance(), maxDistance);
+        for (Car car : this.cars) {
+            maxDistance = Math.max(car.getDistance().getDistance(), maxDistance);
```
stream 의 max 를 사용해볼수도 있겠어요!

```suggestion
    return something.stream()
                .mapToInt(it -> it.getValue())
                .max()
                .getAsInt();
```

ezzanzzan replied:
이전 리팩토링 과정에서 For문을 stream으로 바꾸며, 해당 코드도 stream을 사용할 수 있을까 라는 생각을 해보았는데 어떤 중간연산자를 사용해야 할 지 고민이었는데 코멘트를 보고 stream max 에 대해 찾아보며 적용해 볼 수 있었습니다 !
```
    private int findMaxDistance() {
        return this.cars.stream()
                .mapToInt(car -> car.getDistance().getValue())
                .max()
                .orElse(-1);
    }
```

Gomding replied:
OptionalInt 를 잘 핸들링해주셨군요!
Optional 사용방법에 대해서 추가로 아티클 공유해드릴게요 😃 

http://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/

Gomding commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -2,41 +2,54 @@
+import java.util.Random;
+import java.util.stream.Collectors;
 
 public class Cars {
```
Cars에 대한 테스트도 추가해볼 수 있겠네요 😃

ezzanzzan replied:
Cars에 대한 단위테스트를 하는 부분에서 생성자를 새로 만들어야 한다는 부분에서 고민을 하다 테스트만을 위한 생성자나 메소드의 접근제어자를 private -> public 으로 변경한다는 게 좋지 않다고 생각해서 테스트를 진행하지 못했었습니다.

디엠을 통해 방향점을 찾고 생성자를 추가, 인자를 변경한 뒤
```
class CarsTest {
    Cars cars = new Cars(List.of(new Car("tori", 10), new Car("mango", 20)));

    @Test
    @DisplayName("우승자의 이름을 Return")
    void Should_Winner_When_getWinnerMethod() {
        assertThat(cars.getWinner()).isEqualTo(List.of("mango"));
    }

}
```
와 같이 Cars에 대한 단위테스트를 추가할 수 있었습니다 😊

Gomding reviewed (changes requested):
안녕하세요 토리!!
리팩토링 하느라 고생하셨습니다 🙇 
리팩터링 목록을 문서에 정리까지! 😮 💯 
기존에 원시값들을 다 포장해주신것이 인상깊었습니다
몇가지 코멘트 남겨봤으니 확인 부탁드려요!!

ezzanzzan committed:
docs: 2차 리팩토링 목록 작성

ezzanzzan committed:
refactor: 상수와 객체의 필드 사이 개행 추가

ezzanzzan committed:
refactor: Name, TryCount의 필드 네이밍 value로 변경

ezzanzzan committed:
refactor: forEach, stream 사용 시 개행 리팩토링

ezzanzzan committed:
refactor: Distance 필드 네이밍 value로 변경

ezzanzzan committed:
refactor: findMaxDistance 메소드 for문 대신 stream을 사용하도록 리팩토링

ezzanzzan committed:
refactor: Cars 생성자 추가 및 형식 변경

ezzanzzan committed:
test: Cars 단위 테스트 추가

Gomding commented on src/main/java/racingcar/domain/Cars.java:
```diff
@@ -4,21 +4,23 @@
 
-    public Cars(List<String> carNames) {
-        this.cars = convertToCar(carNames);
+    public Cars(String carNames) {
+        this(Stream.of(carNames.split(DELIMITER))
+                .map(Car::new)
+                .collect(Collectors.toList()));
     }
```
질문주신 부분인데 정적 팩터리 메서드를 사용하면 오버로딩 문제를 해결할 수 있을것같아요!

[정적 팩토리 메서드(Static Factory Method)는 왜 사용할까?](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)

Gomding commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -0,0 +1,18 @@
+    Cars cars = new Cars(List.of(new Car("tori", 10), new Car("mango", 20)));
+
+    @Test
+    @DisplayName("우승자의 이름을 Return")
+    void Should_Winner_When_getWinnerMethod() {
+        assertThat(cars.getWinner()).isEqualTo(List.of("mango"));
+    }
```
`우승자는 여러명이 될 수 있다` 라는 요구사항이 있으니 이부분도 테스트가 있으면 좋을것 같습니다. 😸

Gomding reviewed (approved):
토리 자동차 경주 2단계 진행하느라 고생하셨습니다!!
고민하면서 바꿔주신 부분들에 의해 코드가 많이 개선됐네요 👍 👍 👍 👍 
이번 미션은 이만 merge 하겠습니다.
추가로 코멘트 몇가지 남겼으니 확인하시면 좋을것같습니다~!

남은 과정도 화이팅입니다~!! 🔥 🔥 🔥
